web/packages/teleterm/src/mainProcess/contextMenus/tabContextMenu.ts

```diff
@@ -21,61 +21,151 @@ import {
             {
               label: 'Close',
-              click: () => resolve(TabContextMenuEventType.Close),
+              click: () => resolve({ event: TabContextMenuEventType.Close }),
             },
             {
               label: 'Close Others',
-              click: () => resolve(TabContextMenuEventType.CloseOthers),
+              click: () =>
+                resolve({ event: TabContextMenuEventType.CloseOthers }),
             },
             {
               label: 'Close to the Right',
-              click: () => resolve(TabContextMenuEventType.CloseToRight),
+              click: () =>
+                resolve({ event: TabContextMenuEventType.CloseToRight }),
             },
           ];
         }
 
         function getPtyTemplate(): MenuItemConstructorOptions[] {
           if (
-            options.documentKind === 'doc.terminal_shell' ||
-            options.documentKind === 'doc.terminal_tsh_node'
+            options.document.kind === 'doc.terminal_shell' ||
+            options.document.kind === 'doc.terminal_tsh_node'
           ) {
+            return [
+              {
+                label: 'Duplicate Tab',
+                click: () =>
+                  resolve({ event: TabContextMenuEventType.DuplicatePty }),
+              },
+            ];
+          }
+        }
+
+        function getShellTemplate(): MenuItemConstructorOptions[] {
+          const doc = options.document;
+          if (
+            doc.kind === 'doc.terminal_shell' ||
+            doc.kind === 'doc.gateway_kube'
+          ) {
+            const activeShellId = doc.shellId;
+            const defaultShellId = configService.get('terminal.shell').value;
+            const customShellPath = configService.get(
+              'terminal.customShell'
+            ).value;
+            const customShell =
+              customShellPath && makeCustomShellFromPath(customShellPath);
+            const shellsWithCustom = [...shells, customShell].filter(Boolean);
             return [
               {
                 type: 'separator',
               },
+              ...shellsWithCustom.map(shell => ({
+                label: shell.friendlyName,
+                id: shell.id,
+                type: 'radio' as const,
+                checked: shell.id === activeShellId,
+                click: () => {
+                  // Do nothing when the shell doesn't change.
+                  if (shell.id === activeShellId) {
+                    return;
+                  }
+                  resolve({
+                    event: TabContextMenuEventType.ReopenPtyInShell,
+                    item: shell,
+                  });
+                },
+              })),
               {
-                label: 'Duplicate Tab',
-                click: () => resolve(TabContextMenuEventType.DuplicatePty),
+                label: customShell
+                  ? `Change Custom Shell (${customShell.friendlyName})…`
+                  : 'Select Custom Shell…',
```

Comment:
Do we need the ellipsis? I'm not sure if it's a pattern for context actions that we have anywhere else in the app.

Reply:
Great find!